File: tests/fixtures/gen_jpeg_makernotes_broken.py
"""生成合成 JPEG fixture：nom-exif `parse_exif` 失败但 IFD0 仍可读。

产物：tests/data/sample-jpeg-app1-broken.jpg（一次性，commit 到 git）。

模拟 Canon EOS 7D MakerNotes 偏移异常场景（exiftool 报 `Adjusted MakerNotes
base by -126`）：nom-exif 整体 `parse_exif` 返 Err，丢失全部时间字段。

合成策略：APP1 Exif IFD0 含 Make/Model/ExifIFDPointer；ExifIFD 声称 count=10000
但实际只有 1 entry 的空间，使 nom-exif 在扫 ExifIFD 时越界拒绝；自实现 fallback
通过 `let count = u16_at(...)?` + 后续 `?` 同样越界，但 IFD0 字段已读出。

注意：tidymedia fallback 实际只读 IFD0 + 一个 ExifIFD；ExifIFD 越界时
`scan_ifd` 内 `?` 触发 None → `parse_ifds` 把它当 ExifIFD 子调用失败忽略，
保留 IFD0 字段。所以 IFD0 的 Make/Model + ExifIFD 头一个 entry 的 DTO 都能读。
"""

from __future__ import annotations

import struct
import sys
from pathlib import Path


def _u16le(v: int) -> bytes:
    return struct.pack("<H", v)


def _u32le(v: int) -> bytes:
    return struct.pack("<I", v)


def build_broken_tiff() -> bytes:
    """TIFF: IFD0 含 Make/Model/ExifIFDPointer(→损坏 ExifIFD)。"""
    out = b"II" + _u16le(0x002A) + _u32le(8)
    # IFD0 count=3
    out += _u16le(3)
    # Make @offset 80 (cnt=6: "Cam\0\0\0")
    out += _u16le(0x010F) + _u16le(2) + _u32le(6) + _u32le(80)
    # Model @offset 86 (cnt=7: "Model\0\0")
    out += _u16le(0x0110) + _u16le(2) + _u32le(7) + _u32le(86)
    # ExifIFDPointer @offset 50 (恶意 ExifIFD)
    out += _u16le(0x8769) + _u16le(4) + _u32le(1) + _u32le(50)
    out += _u32le(0)  # IFD0 next-IFD = 0；至此 cum = 8+2+36+4 = 50
    # 恶意 ExifIFD @50：声称 count=10000，但仅放 1 个 DTO entry 的空间
    out += _u16le(10000)  # 50..52
    # DTO entry 52..64：val=93（数据区）
    out += _u16le(0x9003) + _u16le(2) + _u32le(20) + _u32le(93)
    # 填充 64..80 让后续 9998 entries 越界（位于文件末尾外）
    out += b"\x00" * (80 - 64)
    # ASCII 数据
    out += b"Cam\0\0\0"  # 80..86
    out += b"Model\0\0"  # 86..93
    out += b"2017:02:14 10:30:00\0"  # 93..113
    return out


def jpeg_with_app1(tiff: bytes) -> bytes:
    payload = b"Exif\0\0" + tiff
    seg_len = len(payload) + 2  # length 字段含自身 2 字节
    return b"\xff\xd8\xff\xe1" + struct.pack(">H", seg_len) + payload + b"\xff\xd9"


def main() -> int:
    out = Path(__file__).resolve().parents[1] / "data" / "sample-jpeg-app1-broken.jpg"
    out.parent.mkdir(parents=True, exist_ok=True)
    out.write_bytes(jpeg_with_app1(build_broken_tiff()))
    print(f"wrote {out} ({out.stat().st_size} bytes)")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/fixtures/gen_jpeg_makernotes_broken.ts
// 生成合成 JPEG fixture：nom-exif `parse_exif` 失败但 IFD0 仍可读。
//
// 产物：tests/data/sample-jpeg-app1-broken.jpg（一次性，commit 到 git）。
//
// 模拟 Canon EOS 7D MakerNotes 偏移异常场景：nom-exif 整体 `parse_exif` 返 Err。
// APP1 Exif IFD0 含 Make/Model/ExifIFDPointer；ExifIFD 声称 count=10000 但实际
// 只有 1 entry 的空间，使 nom-exif 越界拒绝；自实现 fallback 保留 IFD0 字段。

import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

function u16le(v: number): Buffer {
  const b = Buffer.alloc(2);
  b.writeUInt16LE(v, 0);
  return b;
}

function u32le(v: number): Buffer {
  const b = Buffer.alloc(4);
  b.writeUInt32LE(v >>> 0, 0);
  return b;
}

// TIFF: IFD0 含 Make/Model/ExifIFDPointer(→损坏 ExifIFD)。
function buildBrokenTiff(): Buffer {
  const parts: Buffer[] = [];
  parts.push(Buffer.from("II"), u16le(0x002a), u32le(8));
  parts.push(u16le(3)); // IFD0 count
  parts.push(Buffer.concat([u16le(0x010f), u16le(2), u32le(6), u32le(80)])); // Make
  parts.push(Buffer.concat([u16le(0x0110), u16le(2), u32le(7), u32le(86)])); // Model
  parts.push(Buffer.concat([u16le(0x8769), u16le(4), u32le(1), u32le(50)])); // ExifIFDPointer
  parts.push(u32le(0)); // IFD0 next；至此 cum = 8+2+36+4 = 50
  parts.push(u16le(10000)); // 恶意 ExifIFD @50：count=10000
  parts.push(Buffer.concat([u16le(0x9003), u16le(2), u32le(20), u32le(93)])); // DTO entry
  parts.push(Buffer.alloc(80 - 64)); // 填充，让后续 entries 越界
  parts.push(Buffer.from("Cam\0\0\0")); // 80..86
  parts.push(Buffer.from("Model\0\0")); // 86..93
  parts.push(Buffer.from("2017:02:14 10:30:00\0")); // 93..113
  return Buffer.concat(parts);
}

function jpegWithApp1(tiff: Buffer): Buffer {
  const payload = Buffer.concat([Buffer.from("Exif\0\0"), tiff]);
  const segLen = payload.length + 2; // length 字段含自身 2 字节
  const seg = Buffer.alloc(2);
  seg.writeUInt16BE(segLen, 0);
  return Buffer.concat([Buffer.from([0xff, 0xd8, 0xff, 0xe1]), seg, payload, Buffer.from([0xff, 0xd9])]);
}

const out = join(dirname(fileURLToPath(import.meta.url)), "..", "data", "sample-jpeg-app1-broken.jpg");
mkdirSync(dirname(out), { recursive: true });
const payload = jpegWithApp1(buildBrokenTiff());
writeFileSync(out, payload);
console.log(`wrote ${out} (${payload.length} bytes)`);
